### ISSUE TITLE
Feature parity between OpenGL and Direct3D vo's

### ISSUE THREAD

**Issue by TheFireKahuna** [mpv-player/mpv#2442]
On Windows Direct3D support is miles ahead when it comes to drivers. This often-cited article explains this well:
http://richg42.blogspot.com.au/2014/05/the-truth-on-opengl-driver-quality.html

Additionally, Microsoft doesn't really invest in WGL performance/stability, the focus is on DWM and Direct3D.

Would it be possible to have a Direct3D 11 hardware renderer that is on-par quality-wise with OpenGL? It would be great to have a much more stable presentation in regards to DWM and VSync. If Direct3D support was to be brought into parity, there would be no need for D3D9 support (XP is dying, D3D11 cards are far the norm now and without a D3D11 hardware renderer the alternative would have to be a focus on the DXVA2 API).

Vulcan is coming soon (meaning a shift of resources for driver teams) and Apple has moved its focus away from OpenGL (their new Metal API). Long-term on the Windows platform OpenGL support isn't going to improve. Sure, new versions will be implemented, but refactors or improvements to contexts/windowing isn't going to happen.

**Comment by wm4** [comment]
That's true, OpenGL support on Windows is basically carried by driver vendors, and Microsoft doesn't care the least about it. This has been this way since forever.

Duplicating the vo_opengl render chain for D3D is out of the question, though. It's far too complex. Dropping GL in favor of D3D is also out of the question, because there's no D3D support outside of Windows at all.

There's ANGLE, Google's GLES implementation on top of D3D. This sounds nice in theory, but the featureset provided by GLES/ANGLE is not sufficient to cover any of vo_opengl's advanced features.

Lastly, one could try to make some sort of common API abstraction GL and D3D. It wouldn't have to be a complete abstraction, but only cover enough for vo_opengl to operate, and thus relatively simple. This sounds nice in theory, but I'm sure there would be tons of issues in practice. And certainly not a walk in the park to implement.

So, in summary, I don't see a viable way to change this situation.

**Comment by TheFireKahuna** [comment]
It seems like translation or abstraction would be the only possible solutions then.

ANGLE is working towards OpenGL ES 3.0 support (something around 92% conformance atm), when that is complete would it provide enough scope to allow translation of the OpenGL backend?

**Comment by wm4** [comment]
ES 3.0 is still not good enough.

**Comment by wm4** [comment]
I've looked again. GLES 3.2 actually adds support for high bit depth FBOs. So there may be hope. 3.2 was released 3 months ago, and ANGLE supports not even 3.0 fully, so it's going to take a while.

**Comment by TheFireKahuna** [comment]
Looks like there is hope in the next two years at least.

If https://github.com/mpv-player/mpv/wiki/Stuff-to-do is still part of the plan, that's a great start. If presentation could be implemented through Direct3D, that alone should definitely improve playback. Would opengl-cb be compatible with such an implementation?

**Comment by wm4** [comment]
>If https://github.com/mpv-player/mpv/wiki/Stuff-to-do is still part of the plan, th

It has to be investigated. Nobody knows whether it can be implemented, or whether it works properly.

>Would opengl-cb be compatible with such an implementation?

opengl-cb is for embedding the GL renderer itself in other applications (like Plex Media Player), and does not use any of mpv's windowing code. All it requires is plain OpenGL, and the application has to create the window, OpenGL context, and all the other things.

**Comment by wm4** [comment]
PS: which means that for things like Plex Media Player, which uses Qt, this would have to be implemented in Qt.

**Comment by lamarpavel** [comment]
I would postpone decisions regarding this at least until the Vulkan spec is released for some time. Not because Vulkan is better for this purpose than either OpenGL or Direct3D, it is most likely overkill, but it might turn out to change the situation for driver quality and if libs start to pop up that abstract away lots of boiler plate code, Vulkan might be a better choice for long term support compared to having _two_ renderers (OpenGL + Direct3D).

Note: This is considering that lots of stuff would be re-written anyway, I would not suggest moving away from OpenGL at all right now.

**Comment by wm4** [comment]
Vulkan is a good point; I sure wish it stopped being vaporware.

**Comment by haasn** [comment]
Creating a common abstraction seems unlikely to be easy due to the differences between GLSL and HLSL.

Almost all renderer code would have to be rewritten on top of this abstraction, so it's not much less effort than fully reimplementing it in D3D - which is also out of the question.

**Comment by wm4** [comment]
I wouldn't say it's that bad, but it's most definitely not easy either.

**Comment by wm4** [comment]
I found that recent ANGLE versions already support almost everything what we need.

We added an ANGLE backend, which now has feature parity with real GL, except 10 bit video input, ICC profiles, and nnedi.

Thus there won't be any improvements to the D3D VO. It's rather likely that we remove vo_direct3d entirely. ANGLE supports D3D11 btw., while vo_direct3d is stuck with D3D9.
